File: problem_11/main.py
#!/opt/miniconda3/bin/python


import math
import numpy as np


if __name__ == '__main__':

    in_file = open('in', 'r') 

    row = 20
    col = 20

    product = 1

    grid = []

    for i in range(row):
        line = in_file.readline().strip()
        grid.append(list(map(int, line.split())))

    grid = np.asarray(grid)

    # left and right
    for r in grid:
        for i in range(len(r) - 4):
            if product < math.prod(r[i:i+4]):
                product = math.prod(r[i:i+4])
    
    # rotate the grid
    rotated_grid = grid.transpose()


    # up and down
    for r in rotated_grid:
        for i in range(len(r) - 4):
            if product < math.prod(r[i:i+4]):
                product = math.prod(r[i:i+4])

    # for diagonal elements
    for i in range(-19, 20):
        diag = grid.diagonal(i)
        if diag.size >= 4:
            for j in range(0, diag.size - 4 + 1):
                if np.prod(diag[j:j+4]) > product:
                    product = np.prod(diag[j:j+4])

    grid = np.fliplr(grid)

    # for reverse diagomal elements
    for i in range(-19, 20):
        diag = grid.diagonal(i)
        if diag.size >= 4:
            for j in range(0, diag.size - 4 + 1):
                if np.prod(diag[j:j+4]) > product:
                    product = np.prod(diag[j:j+4])

    print(product)

File: problem_11/main.cpp

#include <iostream>
#include <vector>
using namespace std;

int main() {
    int r = 20, c = 20;
    int product = 1;  
    vector<vector<int> > grid(r, vector<int>(c, 0));

    for (int i = 0; i < r; ++i) {
        for (int j = 0; j < c; ++j) {
            cin >> grid[i][j];
        }
    }

    // left right 

    for (vector<int> : i) {
        for (int j = 0; j < i.size() - 4; ++i) {
            // product of i[0] - i[4] 
            for (int a = j; a < 4; ++a) {
            }
        }
    }

    return 0;
}
